### PR TITLE
C++ SHM

### PR DESCRIPTION
**Pull request by yellowhatter**

### CODE REVIEW

File: zenoh_cpp_vendor/CMakeLists.txt [hunk 1]
```diff
@@ -39,8 +42,11 @@ ament_vendor(zenoh_cpp_vendor
   VCS_URL https://github.com/eclipse-zenoh/zenoh-cpp
   VCS_VERSION e8eca99b4eaff0963e52aefd8405909c1552080d
   CMAKE_ARGS
-    -DZENOHCXX_ZENOHC=OFF
-)
+      -DZENOHCXX_ZENOHC=ON
+      -DZENOHCXX_ZENOHPICO=OFF
+      -DZENOHC_BUILD_WITH_SHARED_MEMORY:BOOL=${RMW_ZENOH_BUILD_WITH_SHARED_MEMORY}
+      -DZENOHC_BUILD_WITH_UNSTABLE_API=TRUE
+  )
```
Comment (YuanYuYuan) [inline zenoh_cpp_vendor/CMakeLists.txt]: Why don't we build zenoh-c with SHM? zenoh-cpp does not contain a flag for `ZENOHC_BUILD_WITH_SHARED_MEMORY` nor `ZENOHC_BUILD_WITH_UNSTABLE_API`.

File: rmw_zenoh_cpp/src/detail/shm_context.hpp [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+#ifndef DETAIL__SHM_CONTEXT_HPP_
+#define DETAIL__SHM_CONTEXT_HPP_
+
+#include <zenoh.hxx>
+
+namespace rmw_zenoh_cpp
+{
+///=============================================================================
+#ifdef RMW_ZENOH_BUILD_WITH_SHARED_MEMORY
```
Comment (YuanYuYuan) [inline rmw_zenoh_cpp/src/detail/shm_context.hpp]: Can we wrap the whole content in this header file with `#ifdef RMW_ZENOH_BUILD_WITH_SHARED_MEMORY`?

File: rmw_zenoh_cpp/src/detail/shm_context.cpp [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+// limitations under the License.
+
+#include <stdexcept>
+
+#include "shm_context.hpp"
+
+namespace rmw_zenoh_cpp
+{
+///=============================================================================
+#ifdef RMW_ZENOH_BUILD_WITH_SHARED_MEMORY
```
Comment (YuanYuYuan) [inline rmw_zenoh_cpp/src/detail/shm_context.cpp]: Can we wrap the whole content in this header file with `#ifdef RMW_ZENOH_BUILD_WITH_SHARED_MEMORY`?

File: rmw_zenoh_cpp/src/detail/shm_context.cpp [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+ShmContext::ShmContext(size_t alloc_size, size_t msgsize_threshold)
+:  // Create Layout for provider's memory
+   // Provider's alignment will be 1 byte as we are going to make only 1-byte aligned allocations
```
Comment (YuanYuYuan) [inline rmw_zenoh_cpp/src/detail/shm_context.cpp]: Explain a little bit about the `AllocAlignment`, e.g. "will be 1 bytes" -> "will be 1 (=2^0) bytes"

File: rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp [hunk 5]
```diff
@@ -179,21 +179,21 @@ class rmw_context_impl_s::Data final
           "rmw_zenoh_cpp", "[rmw_context_impl_s] z_call received an invalid reply.\n");
       }
     }
-
-    // Initialize the shm manager if shared_memory is enabled in the config.
-    shm_provider_ = std::nullopt;
-#ifndef _MSC_VER
```
Comment (YuanYuYuan) [inline rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp:171]: Should we retain the Windows check? Does the SHM feature work on Windows?

**Review by YuanYuYuan — CHANGES_REQUESTED** [review]
In addition to the comments I left, I'd like to discuss on some points.

1. We allow the users to build rmw_zenoh_cpp without SHM feature. What if a user enables the SHM feature in the config that is the default behavior and builds the rmw_zenoh_cpp without SHM support? It seems not to report any error.
2. Can you change the base branch to the rolling? The upstream has been ported to zenoh-cpp.

File: zenoh_cpp_vendor/CMakeLists.txt (hunk 1 above)
Comment (yellowhatter) [reply zenoh_cpp_vendor/CMakeLists.txt]: right, removed

File: rmw_zenoh_cpp/src/detail/shm_context.hpp (hunk 2 above)
Comment (yellowhatter) [reply rmw_zenoh_cpp/src/detail/shm_context.hpp]: done

File: rmw_zenoh_cpp/src/detail/shm_context.cpp (hunk 3 above)
Comment (yellowhatter) [reply rmw_zenoh_cpp/src/detail/shm_context.cpp]: done

File: rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp (hunk 5 above)
Comment (yellowhatter) [reply rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp:171]: SHM for Windows is supported and works :)

File: rmw_zenoh_cpp/src/detail/shm_context.cpp (hunk 4 above)
Comment (yellowhatter) [reply rmw_zenoh_cpp/src/detail/shm_context.cpp]: done

File: rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp (hunk 5 above)
Comment (YuanYuYuan) [reply rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp:171]: Okay. Hope we can pass all the tests on Windows. :smiley:

File: rmw_zenoh_cpp/src/detail/shm_context.cpp [hunk 6]
```diff
@@ -0,0 +1,33 @@
+// Copyright 2023 Open Source Robotics Foundation, Inc.
```
Comment (evshary) [inline rmw_zenoh_cpp/src/detail/shm_context.cpp]: Perhaps this should be 2024.

File: rmw_zenoh_cpp/src/detail/shm_context.hpp [hunk 7]
```diff
@@ -0,0 +1,36 @@
+// Copyright 2023 Open Source Robotics Foundation, Inc.
```
Comment (evshary) [inline rmw_zenoh_cpp/src/detail/shm_context.hpp]: Same here

**Review by evshary — APPROVED** [review]
LGTM, but since we have `RMW_ZENOH_BUILD_WITH_SHARED_MEMORY` to choose whether we want to build with shared memory, it might be good to test it with CI.

File: rmw_zenoh_cpp/src/detail/shm_context.hpp (hunk 7 above)
Comment (yellowhatter) [reply rmw_zenoh_cpp/src/detail/shm_context.hpp]: fixed

File: rmw_zenoh_cpp/src/detail/shm_context.cpp (hunk 6 above)
Comment (yellowhatter) [reply rmw_zenoh_cpp/src/detail/shm_context.cpp]: fixed

**Comment by yellowhatter** [comment]
> LGTM, but since we have `RMW_ZENOH_BUILD_WITH_SHARED_MEMORY` to choose whether we want to build with shared memory, it might be good to test it with CI.

Right: https://github.com/ZettaScaleLabs/rmw_zenoh/issues/50

**Comment by yellowhatter** [comment]
> In addition to the comments I left, I'd like to discuss on some points.
> 
> 1. We allow the users to build rmw_zenoh_cpp without SHM feature. What if a user enables the SHM feature in the config that is the default behavior and builds the rmw_zenoh_cpp without SHM support? It seems not to report any error.
> 2. Can you change the base branch to the rolling? The upstream has been ported to zenoh-cpp.

2. done

1. Well, this question is more related to Zenoh-Rust implementation rather than RMW. If there is no SHM feature enabled, the SHM section is just ignored. Attempts to have some warnings for combination of SHM-Config + Non-SHM build will just complexify the setup for nothing.

**Pull request merged**